crates/ingress-client/src/session.rs

```diff
@@ -0,0 +1,501 @@
+}
+
+impl SessionHandle {
+    /// Enqueues an ingest request along with the owned permit and returns a future tracking commit outcome.
+    pub fn ingest(
+        &self,
+        permit: OwnedSemaphorePermit,
+        record: IngestRecord,
+    ) -> Result<RecordCommit, SessionClosed> {
+        let (commit, resolver) = RecordCommit::new();
+        self.tx
+            .send((
+                RecordTracker {
+                    _permit: permit,
+                    resolver,
+                },
+                record,
+            ))
+            .map_err(|_| SessionClosed)?;
+
+        Ok(commit)
+    }
+}
+
+enum SessionState {
+    Connecting { retry: RetryIter<'static> },
+    Connected { connection: Connection },
+    Disconnected,
+    Shutdown,
+}
+
+/// Background task that drives the lifecycle of a single partition connection.
+pub struct PartitionSession<T> {
+    partition: PartitionId,
+    partition_routing: PartitionRouting,
+    networking: Networking<T>,
+    opts: SessionOptions,
+    rx: UnboundedReceiverStream<(RecordTracker, IngestRecord)>,
+    tx: mpsc::UnboundedSender<(RecordTracker, IngestRecord)>,
+    inflight: VecDeque<IngressBatch>,
+}
+
+impl<T> PartitionSession<T> {
+    fn new(
+        networking: Networking<T>,
+        partition_routing: PartitionRouting,
+        partition: PartitionId,
+        opts: SessionOptions,
+    ) -> Self {
+        let (tx, rx) = mpsc::unbounded_channel();
+        let rx = UnboundedReceiverStream::new(rx);
+
+        Self {
+            partition,
+            partition_routing,
+            networking,
+            opts,
+            inflight: Default::default(),
+            rx,
+            tx,
+        }
+    }
+
+    /// Returns a handle that can be used by callers to enqueue new records.
+    pub fn handle(&self) -> SessionHandle {
+        SessionHandle {
+            tx: self.tx.clone(),
+        }
+    }
+}
+
+impl<T> PartitionSession<T>
+where
+    T: TransportConnect,
+{
+    /// Runs the session state machine until shut down, reacting to cancellation and connection errors.
+    pub async fn start(mut self, ctx: CancellationToken) {
+        let mut state = SessionState::Connecting {
+            retry: self.opts.connect_retry_policy.clone().into_iter(),
+        };
+
+        debug!(
+            partition_id = %self.partition,
+            "Starting ingress partition session",
+        );
+
+        loop {
+            state = match state {
+                SessionState::Connecting { retry } => self.connect(&ctx, retry).await,
+                SessionState::Connected { connection } => self.connected(&ctx, connection).await,
+                SessionState::Disconnected => SessionState::Connecting {
+                    retry: self.opts.connect_retry_policy.clone().into_iter(),
+                },
+                SessionState::Shutdown => {
+                    self.rx.close();
+                    break;
+                }
+            }
```

Comment:
Instead of passing down the `CancellationToken` into every method, we could also do `CancellationToken::run_until_cancelled`. This would have the advantage that we don't need to add cancellation to every await point.